File: qt-src/interface/graphicboard.cpp
#include "graphicboard.h"
#include "logger.h"
#include <QInputDialog>
#include "gamemanager.h"
#include <QStringList>
#include <lang/language.h>
#include <lang/englishlang.h>
#include <lang/russianlang.h>

GraphicBoard::GraphicBoard(int width, int height, QWidget *parent) :
    width(width),
    height(height),
    QWidget(parent)
{
    Language *language;
    QStringList langList;
    langList.push_back("Russian");
    langList.push_back("English");
    QString gameLang = QInputDialog::getItem(this, tr("Choose language of the game"), tr("Выберите язык игры"), langList);
    if (gameLang == "Russian"){
        language = new RussianLang;
    }
    else{
        language = new EnglishLang;
    }
    tableLayout = new QGridLayout;
    for (int i = 0; i < height; ++i) {
        buttons.push_back(QVector<ButtonCell*>(width));
        for (int j = 0; j < width; ++j) {
            buttons[i][j] = new ButtonCell(language);
            buttons[i][j]->connectToPanel(this);
            tableLayout->addWidget(buttons[i][j], i, j);
        }
    }
    gamePanel = new QHBoxLayout(this);
    gameBoardPanel = new QVBoxLayout;
    gameBoardPanel->addLayout(tableLayout);
    wordPanel = new QHBoxLayout;
    currentWord = new QLabel("");
    commitButton = new QPushButton("Enter!");
    commitGiveUp = new QPushButton("Give Up");
    wordPanel->addWidget(commitGiveUp);
    wordPanel->addWidget(currentWord);
    wordPanel->addWidget(commitButton);
    gameBoardPanel->addLayout(wordPanel);


    firstPlayerScore = new QLabel(tr("0"), this);
    firstPlayerWords = new QListWidget(this);
    firstPlayerPanel = new QVBoxLayout(this);
    firstPlayerPanel->addWidget(firstPlayerScore);
    firstPlayerPanel->addWidget(firstPlayerWords);

    secondPlayerScore = new QLabel(tr("0"), this);
    secondPlayerWords = new QListWidget(this);
    secondPlayerPanel = new QVBoxLayout(this);
    secondPlayerPanel->addWidget(secondPlayerScore);
    secondPlayerPanel->addWidget(secondPlayerWords);

    gamePanel->addLayout(firstPlayerPanel);
    gamePanel->addLayout(gameBoardPanel);
    gamePanel->addLayout(secondPlayerPanel);

    setLayout(gamePanel);

    int players = QInputDialog::getInt(this, tr("Enter number of players"), tr("Введите число игроков"), 1, 1, 2);
    if (players == 2) {
        gameManager = new GameManager(language, width, height, players);
    } else {
        QStringList list;
        list.push_back(tr("EASY"));
        list.push_back("MEDIUM");
        list.push_back(tr("HARD"));
        list.push_back("HARDEST");
        QString level = QInputDialog::getItem(this, tr("Choose level"), tr("Выберите сложность"), list);
        gameManager = new GameManager(language, width, height, players, level);
    }
    connectToPlayers(gameManager->getFirstPlayer(), gameManager->getSecondPlayer());
    gameManager->getFirstPlayer()->connectToInterface(this);
    gameManager->getSecondPlayer()->connectToInterface(this);
    QString word = gameManager->getFirstWord();
    for (int i = 0; i < width; ++i) {
        buttons[height/2][i]->setText(QString(word[i]));
    }

    for (int i = 0; i < width; ++i) {
        buttons[height/2][i]->setMenu(NULL);
    }

    connect(commitButton, SIGNAL(clicked()), this, SLOT(onCommitButtonClicked()));
    connect(commitGiveUp, SIGNAL(clicked()), this, SLOT(onCommitGiveUpClicked()));


    connect(gameManager, SIGNAL(gameEnded(const QString&)),
            this, SLOT(finishGame(const QString&)));
    connect(this, SIGNAL(quit()), parent, SLOT(resetField()));
}

void GraphicBoard::connectToPlayers(Player *player1, Player* player2)
{
    connect(this, SIGNAL(chooseLetterFirst(QPair<QPair<int,int>, QChar>)),
            player1, SLOT(onLetterChosen(QPair<QPair<int,int>,QChar>)));
    connect(this, SIGNAL(chooseLetterSecond(QPair<QPair<int,int>, QChar>)),
            player2, SLOT(onLetterChosen(QPair<QPair<int,int>,QChar>)));
    connect(this, SIGNAL(pushLetterFirst(QPair<int,int>)),
            player1, SLOT(onLetterPushed(QPair<int,int>)));
    connect(this, SIGNAL(pushLetterSecond(QPair<int,int>)),
            player2, SLOT(onLetterPushed(QPair<int,int>)));
    connect(this, SIGNAL(commitWordFirst()), player1, SLOT(onWordCommited()));
    connect(this, SIGNAL(commitWordSecond()), player2, SLOT(onWordCommited()));
    connect(this, SIGNAL(giveUpFirst()), player1, SLOT(sendEnd()));
    connect(this, SIGNAL(giveUpSecond()), player2, SLOT(sendEnd()));

}


void GraphicBoard::runStep()
{
    gameManager->runGame();
}


void GraphicBoard::onCellPushed()
{
    int xSender = 0;
    int ySender = 0;
    for (int i = 0; i < height; ++i) {
        for (int j = 0; j < width; ++j) {
            if (reinterpret_cast<ButtonCell*>(sender()) == buttons[i][j]) {
                xSender = i;
                ySender = j;
            }
        }
    }
    //  currentWord->setText(currentWord->text() + buttons[xSender][ySender]->text());
    if (gameManager->getCurrentPlayer() == 1) {
        emit pushLetterFirst(qMakePair(xSender, ySender));
    } else {
        emit pushLetterSecond(qMakePair(xSender, ySender));
    }
}

void GraphicBoard::onCellChosen(QChar letter)
{
    int xSender = 0;
    int ySender = 0;
    for (int i = 0; i < height; ++i) {
        for (int j = 0; j < width; ++j) {
            if (reinterpret_cast<ButtonCell*>(sender()) == buttons[i][j]) {
                xSender = i;
                ySender = j;
            }
        }
    }
    QPair<QPair<int,int>, QChar> sendLetter = qMakePair(qMakePair(xSender, ySender), letter);
    if (gameManager->getCurrentPlayer() == 1) {
        emit chooseLetterFirst(sendLetter);
    } else {
        emit chooseLetterSecond(sendLetter);
    }
}

void GraphicBoard::afterCellPushed(QPair<int, int>coordinates)
{
    currentWord->setText(currentWord->text() + buttons[coordinates.first][coordinates.second]->text());
}

void GraphicBoard::afterCellChoosen(QPair<QPair<int, int>, QChar> coordinates)
{
    buttons[coordinates.first.first][coordinates.first.second]->setMenu(NULL);
    buttons[coordinates.first.first][coordinates.first.second]->setText(QString(coordinates.second));
}

void GraphicBoard::afterWordCommited(QString word) {
    currentWord->setText(NULL);
    if (sender() == gameManager->getFirstPlayer()) {
        firstPlayerWords->addItem(word);
    } else {
        secondPlayerWords->addItem(word);
    }
    if (sender() == gameManager->getFirstPlayer()) {
        int score = gameManager->getFirstPlayer()->getScore();
        firstPlayerScore->setText(QString::number(score));
    } else {
        int score = gameManager->getSecondPlayer()->getScore();
        secondPlayerScore->setText(QString::number(score));
    }
    runStep();
}

void GraphicBoard::onPlayerResetWord(const QPair<int, int> &coordinates)
{
    buttons[coordinates.first][coordinates.second]->createMenu();
    buttons[coordinates.first][coordinates.second]->setText(QString(""));
    currentWord->setText(QString(""));
}

void GraphicBoard::onCommitButtonClicked()
{
    if (gameManager->getCurrentPlayer() == 1) {
        emit commitWordFirst();
    } else {
        emit commitWordSecond();
    }
}

void GraphicBoard::onCommitGiveUpClicked()
{
    if (gameManager->getCurrentPlayer() == 1) {
        emit giveUpFirst();
    } else {
        emit giveUpSecond();
    }
}

void GraphicBoard::finishGame(const QString &message)
{
    QMessageBox box(this);
    box.setText(message);
    QPushButton* okButton = box.addButton(QMessageBox::Ok);
    box.exec();
    if (box.clickedButton() == okButton) {
        emit quit();
    }
}

File: python-src/interface/GraphicBoard.py
from PySide import QtGui, QtCore
from GameManager import GameManager
from Letter import Coordinates, CellLetter
from Player import Player
from interface.ButtonCell import ButtonCell
from lang.EnglishLanguage import EnglishLanguage
from lang.RussianLanguage import RussianLanguage

__author__ = 'akhtyamovpavel'

class GraphicBoard(QtGui.QWidget):

    push_letter_first = QtCore.Signal(Coordinates)
    push_letter_second = QtCore.Signal(Coordinates)
    choose_letter_first = QtCore.Signal(CellLetter)
    choose_letter_second = QtCore.Signal(CellLetter)
    commit_word_first = QtCore.Signal()
    commit_word_second = QtCore.Signal()
    quit = QtCore.Signal()
    give_up_first = QtCore.Signal()
    give_up_second = QtCore.Signal()
    
    def __init__(self, width, height, parent):
        super(GraphicBoard, self).__init__()

        self.__width__ = width
        self.__height__ = height
        self.__language__ = None
        lang_list = ['Russian', 'English']
        game_lang, flag2 = QtGui.QInputDialog.getItem(self, 'Choose language of the game', 'Выберите язык игры', lang_list)
        if game_lang == 'Russian':
            self.__language__ = RussianLanguage()
        else:
            self.__language__ = EnglishLanguage()

        self.table_layout = QtGui.QGridLayout()
        self.buttons = list()

        for i in range(height):
            tmp_button_list = list()
            for j in range(width):
                tmp_button = ButtonCell(self.__language__)
                tmp_button.connect_to_panel(self)
                self.table_layout.addWidget(tmp_button, i, j)
                tmp_button_list.append(tmp_button)
            self.buttons.append(tmp_button_list)

        self.game_panel = QtGui.QHBoxLayout()
        self.game_board_panel = QtGui.QVBoxLayout()
        self.game_board_panel.addLayout(self.table_layout)

        self.word_panel = QtGui.QHBoxLayout()
        self.current_word = QtGui.QLabel('')
        self.commit_button = QtGui.QPushButton('Enter!')
        self.commit_give_up = QtGui.QPushButton('Give up')
        self.word_panel.addWidget(self.commit_give_up)
        self.word_panel.addWidget(self.current_word)
        self.word_panel.addWidget(self.commit_button)

        self.game_board_panel.addLayout(self.word_panel)

        self.first_player_score = QtGui.QLabel('0', self)
        self.first_player_words = QtGui.QListWidget(self)
        self.first_player_panel = QtGui.QVBoxLayout()
        self.first_player_panel.addWidget(self.first_player_score)
        self.first_player_panel.addWidget(self.first_player_words)

        self.second_player_score = QtGui.QLabel('0', self)
        self.second_player_words = QtGui.QListWidget(self)
        self.second_player_panel = QtGui.QVBoxLayout()
        self.second_player_panel.addWidget(self.second_player_score)
        self.second_player_panel.addWidget(self.second_player_words)

        self.game_panel.addLayout(self.first_player_panel)
        self.game_panel.addLayout(self.game_board_panel)
        self.game_panel.addLayout(self.second_player_panel)

        self.setLayout(self.game_panel)

        players, flag = QtGui.QInputDialog.getInt(self, 'Enter number of players', 'Введите число игроков', 1, 1, 2)
        if players == 2:
            self.__game_manager__ = GameManager(self.__language__, width=width, height=height, players_number=players)
        else:
            list_levels = ['EASY', 'MEDIUM', 'HARD', 'HARDEST']
            level = QtGui.QInputDialog.getItem(self, 'Choose level', 'Выберите сложность', list_levels)
            self.__game_manager__ = GameManager(self.__language__, width, height, players, level[0])

        self.connect_to_players(self.__game_manager__.get_first_player(),
                                self.__game_manager__.get_second_player())

        self.__game_manager__.get_first_player().connect_to_interface(self)
        self.__game_manager__.get_second_player().connect_to_interface(self)

        word = self.__game_manager__.get_first_word()
        for i in range(width):
            self.buttons[height//2][i].setText(word[i])
            self.buttons[height//2][i].setMenu(None)

        self.commit_button.clicked.connect(self.on_commit_button_clicked)
        self.commit_give_up.clicked.connect(self.on_commit_give_up_clicked)

        self.__game_manager__.game_ended.connect(self.finish_game)


        self.quit.connect(parent.reset_field)


    @QtCore.Slot()
    def on_cell_pushed(self):
        x_sender = 0
        y_sender = 0
        for i in range(self.__height__):
            for j in range(self.__width__):
                if self.sender() == self.buttons[i][j]:
                    x_sender = i
                    y_sender = j
        send_letter = Coordinates(x_sender, y_sender)
        if self.__game_manager__.get_current_player() == 1:
            self.push_letter_first.emit(send_letter)
        else:
            self.push_letter_second.emit(send_letter)

    @QtCore.Slot(str)
    def on_cell_chosen(self, letter):
        x_sender = 0
        y_sender = 0
        for i in range(self.__height__):
            for j in range(self.__width__):
                if self.sender() == self.buttons[i][j]:
                    x_sender = i
                    y_sender = j

        send_letter = CellLetter(x_sender, y_sender, letter)
        if self.__game_manager__.get_current_player() == 1:
            self.choose_letter_first.emit(send_letter)
        else:
            self.choose_letter_second.emit(send_letter)

    @QtCore.Slot(Coordinates)
    def after_cell_pushed(self, coordinates: Coordinates):
        self.current_word.setText(self.current_word.text() +
                                      self.buttons[coordinates.x][coordinates.y].text())

    @QtCore.Slot(CellLetter)
    def after_cell_chosen(self, coordinates: CellLetter):
        self.buttons[coordinates.x][coordinates.y].setMenu(None)
        self.buttons[coordinates.x][coordinates.y].setText(coordinates.letter)


    @QtCore.Slot(str)
    def after_word_committed(self, word):
        self.current_word.setText(None)
        if self.sender() == self.__game_manager__.get_first_player():
            self.first_player_words.addItem(word)
            score = self.__game_manager__.get_first_player().get_score()
            self.first_player_score.setText(str(score))
        else:
            self.second_player_words.addItem(word)
            score = self.__game_manager__.get_second_player().get_score()
            self.second_player_score.setText(str(score))
        self.run_step()


    @QtCore.Slot(Coordinates)
    def on_player_reset_word(self, coordinates: Coordinates):
        self.buttons[coordinates.x][coordinates.y].create_menu()
        self.buttons[coordinates.x][coordinates.y].setText('')
        self.current_word.setText('')

    @QtCore.Slot()
    def on_commit_button_clicked(self):
        if self.__game_manager__.get_current_player() == 1:
            self.commit_word_first.emit()
        else:
            self.commit_word_second.emit()

    @QtCore.Slot()
    def on_commit_give_up_clicked(self):
        if self.__game_manager__.get_current_player() == 1:
            self.give_up_first.emit()
        else:
            self.give_up_second.emit()

    @QtCore.Slot(str)
    def finish_game(self, message):
        box = QtGui.QMessageBox(self)
        box.setText(message)
        ok_button = box.addButton(QtGui.QMessageBox.Ok)
        box.exec_()
        if box.clickedButton() == ok_button:
            self.quit.emit()

    def connect_to_players(self, player1: Player, player2: Player):
        self.choose_letter_first.connect(player1.on_letter_chosen)
        self.push_letter_first.connect(player1.on_letter_pushed)

        self.choose_letter_second.connect(player2.on_letter_chosen)
        self.push_letter_second.connect(player2.on_letter_pushed)

        self.commit_word_first.connect(player1.on_word_committed)
        self.commit_word_second.connect(player2.on_word_committed)

        self.give_up_first.connect(player1.send_end)
        self.give_up_second.connect(player2.send_end)


    def run_step(self):
        self.__game_manager__.run_game()



